mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -2848,6 +2848,41 @@ struct LLVMOpAsmDialectInterface : public OpAsmDialectInterface {
 // DialectInlinerInterface
 //===----------------------------------------------------------------------===//
 
+[[nodiscard]] static bool allocaHasConstantSize(LLVM::AllocaOp allocaOp) {
+  Operation *definingOp = allocaOp.getArraySize().getDefiningOp();
+  return definingOp && definingOp->hasTrait<OpTrait::ConstantLike>();
+}
+
+/// Move all `LLVM::AllocaOp` operations with a constant size in `sourceBlock`
+/// to the end of `entryBlock`. If necessary, the operation defining the
+/// constant will be moved as well.
+static void moveConstantAllocasToEntryBlock(Block *sourceBlock,
+                                            Block *entryBlock) {
+  assert(entryBlock->isEntryBlock() &&
+         "Second argument must be an entry block.");
+  assert(sourceBlock->getParent() == entryBlock->getParent() &&
+         "Blocks must belong to the same region.");
+  if (sourceBlock == entryBlock)
+    // Nothing to do.
+    return;
+  SmallVector<LLVM::AllocaOp> allocasToMove;
+  // Don't walk nested regions of the block, since we can't be sure of their
+  // semantics.
+  for (auto allocaOp : sourceBlock->getOps<LLVM::AllocaOp>()) {
+    if (!allocaHasConstantSize(allocaOp))
+      continue;
+    allocasToMove.push_back(allocaOp);
+  }
+  for (auto allocaOp : allocasToMove) {
+    allocaOp->moveBefore(&entryBlock->back());
```

Comment:
entryBlock->back() is the terminator of the block I assume?

Reply:
Yeah, I'll change this to `moveBefore(entryBlock->getTerminator())`